Handle EINTR from send/recv/poll/select to try again as the error is not fatal

sodabrew opened the pull request:
Fixes #955 

Would it make sense to handle EINTR in the wrapper macros `BLOCK_ADJUST` and `BLOCK_ADJUST_ERRNO` instead?

**macOS**:
`man select`:
> select() returns with an error, including one due to an interrupted call, the descriptor sets will be unmodified and the global variable errno will be set to indicate the error.

`man poll`:
> If poll() returns with an error, including one due to an interrupted call, the fds array will be unmodified and the global variable errno will be set to indicate the error.

`man recv`:
> [EINTR]            The receive was interrupted by delivery of a signal before any data were available.

`man send`:
> [EINTR]            A signal interrupts the system call before any data is transmitted.

**Linux**:
`man select`:
> EINTR The function was interrupted before any of the selected events occurred and before the timeout interval expired.

`man poll`:
> EINTR A signal was caught during poll().
Not super helpful, but the text implies behavior as if no other actions were taken.

`man recv`:
> EINTR The receive was interrupted by delivery of a signal before any data were available
`man send`:
>EINTR A signal interrupted send() before any data was transmitted.

**FreeBSD**:
Does not mention EINTR at all in `send` but for `recv`, `poll`, `select` the text follow as macOS and Linux.
https://man.freebsd.org/cgi/man.cgi?sektion=2&query=send

**Windows**:
It looks like `WSAEINTR` should be handled identically as above, ~however that is NOT implemented in this PR (yet! feedback appreciated!)~ Update: added Windows handling, but did not adapt a test for it.
: https://learn.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2

vszakats commented:
checksrc needs fixing indentation and long comments (for all 3 hunks):
```
./src/misc.c:167:86: warning: Longer than 79 columns (LONGLINE)
         /* Profiling tools that use SIGPROF can cause EINTR responses. For recv, this
./src/misc.c:168:96: warning: Longer than 79 columns (LONGLINE)
          * is defined as not returning any data, but there may be data waiting so call again */
./src/misc.c:170:11: warning: not indented 4 steps (uses 2) (INDENTATION)
           return 0;
           ^
```

sodabrew commented:
I fixed the indentation and added support for Windows WSA errors in src/session.c, please let me know if this is too invasive a change.

Other direct uses of errno in src/misc.c are wrapped for win32 with wsa2errno, which is static to that file.

What's the difference in use cases between libssh2_poll and _libssh2_wait_socket? Asking because I wonder if _libssh2_wait_socket could move to misc.c where there are some other socket-related and session-struct-argument but non-public-api (start with underscore) functions, along with wsa2errno()

The two functions are very different in terms of how they're handling the FDs arrays for poll/select, so there wasn't any obvious way to make one a convenience wrapper around the other. Notably, libssh2_poll only uses the return values from poll/select if greater-than-zero, so error conditions seem to be ignored. (I suppose that's as good a reason as any to mark this deprecated? https://www.libssh2.org/libssh2_poll.html)

sodabrew commented:
Added tests based on the example from #955 to show `master` branch fails then this branch solves for EINTR.
```
On master:
      Start 25: test_sshd
25/43 Test #25: test_sshd .................................***Failed    0.33 sec
On branch:
      Start 25: test_sshd
25/43 Test #25: test_sshd .................................   Passed    0.41 sec
```

sodabrew commented:
Thanks for approving the test run, found I can run `./ci/checksrc.sh` locally pre-check my fixes :) Pushed again to pass the linter!

vszakats commented on src/misc.c:
```diff
@@ -159,23 +159,28 @@ _libssh2_recv(libssh2_socket_t sock, void *buffer, size_t length,
+    if(rc < 0) {
 #ifdef WIN32
-    if(rc < 0)
-        return -wsa2errno();
+        int err = _libssh2_wsa2errno();
 #else
-    if(rc < 0) {
+        int err = errno;
+#endif
```
I suggest deduping the `int err` declration into the top of the block. Also in the other similar cased in the patch:
```
... {
  int err;
#ifdef WIN32
  err = ...
#else
  err = ...
#endif
```

sodabrew replied:
Updated

vszakats commented on src/misc.c:
```diff
@@ -192,18 +197,23 @@ _libssh2_send(libssh2_socket_t sock, const void *buffer, size_t length,
+        int err = _libssh2_wsa2errno();
 #else
-    if(rc < 0) {
+        int err = errno;
```
Here also.

vszakats commented on src/session.c:
```diff
@@ -668,6 +668,17 @@ int _libssh2_wait_socket(LIBSSH2_SESSION *session, time_t start_time)
+        int err = _libssh2_wsa2errno();
+#else
+        int err = errno;
+#endif
```
Here also.

vszakats commented on tests/test_ssh2.c:
```diff
@@ -48,6 +78,24 @@ int main(int argc, char *argv[])
+    pthread_t thread;
+
+    struct sigaction signal_handler_config = {
+      .sa_flags = SA_RESTART | SA_SIGINFO,
+      .sa_sigaction = signal_handler_function
```
```suggestion
        .sa_flags = SA_RESTART | SA_SIGINFO,
        .sa_sigaction = signal_handler_function
```
```suggestion
      .sa_flags = SA_RESTART | SA_SIGINFO,
      .sa_sigaction = signal_handler_function
```

sodabrew replied:
Pushed with four spaces

vszakats commented on tests/test_ssh2.c:
```diff
@@ -35,6 +40,31 @@ static void portable_sleep(unsigned int seconds)
+
+static void signal_handler_function(int a, siginfo_t *b, void *c)
+{
+    /* Prevent unused variable warnings */
```
```suggestion
```

sodabrew commented on tests/test_ssh2.c:
```diff
@@ -35,6 +40,31 @@ static void portable_sleep(unsigned int seconds)
+static void *background_thread(void *unused)
+{
+    fprintf(stderr, "Profiling thread sending SIGPROF every 1ms!\n");
+    (void)unused;
```
done

sodabrew commented on src/misc.c:
```diff
@@ -159,23 +159,29 @@ _libssh2_recv(libssh2_socket_t sock, void *buffer, size_t length,
+           recv() does not modify its arguments when it returns EINTR,
+           but there may be data waiting, so the caller should try again */
+        if(err == EINTR)
+            return 0;
```
I observed in tests that returning `-EAGAIN` here also passes. That might be more consistent with this block of code.

sodabrew commented on tests/test_ssh2.c:
```diff
@@ -48,6 +78,24 @@ int main(int argc, char *argv[])
+    main_thread = pthread_self();
+
+    pthread_create(&thread, NULL, background_thread, NULL);
+#endif
```
Updated to test for pthread.h / signal.h

sodabrew commented:
👋 friendly nudge, please let me know if there's anything I should adjust or things to discuss for this PR? thank you!

vszakats commented:
Aside from the smaller issues. My main concern is adding pthreads as a dependency to libssh2. I'm personally not ready to take this burden.

Any input from fellow maintainers?

sodabrew commented:
That’s fair, though I’ll say that the pthread library is only used in the
tests, and it should fail gracefully if not present (not checking the eintr
functionality in that run)

I could redo it used fork() if that would resolve this concern?

vszakats commented on src/session.c:
```diff
@@ -670,6 +670,18 @@ int _libssh2_wait_socket(LIBSSH2_SESSION *session, time_t start_time)
     }
     if(rc < 0) {
+        int err;
+#ifdef WIN32
```
```suggestion
#ifdef _WIN32
```

vszakats commented on src/misc.h:
```diff
@@ -79,6 +79,11 @@ int _libssh2_error_flags(LIBSSH2_SESSION* session, int errcode,
                          const char *errmsg, int errflags);
 int _libssh2_error(LIBSSH2_SESSION* session, int errcode, const char *errmsg);
 
+#ifdef WIN32
```
```suggestion
#ifdef _WIN32
```

vszakats commented on tests/test_ssh2.c:
```diff
@@ -40,6 +45,30 @@ static void portable_sleep(unsigned int seconds)
 #endif
 }
 
+#if defined(HAVE_PTHREAD_H) && defined(HAVE_SIGNAL_H)
+static pthread_t main_thread;
+
+static void *background_thread(void *unused)
```
```suggestion
#if defined(HAVE_PTHREAD_H)
static void *background_thread(void *unused)
```

vszakats commented on tests/test_ssh2.c:
```diff
@@ -24,6 +24,11 @@
 #include <stdio.h>
 #include <stdlib.h>  /* for getenv() */
 
+#if defined(HAVE_PTHREAD_H) && defined(HAVE_SIGNAL_H)
+#include <pthread.h>
+#include <signal.h>
+#endif
```
```suggestion
#if defined(HAVE_PTHREAD_H)
#include <pthread.h>
#include <signal.h>

static pthread_t main_thread;
#endif
```
`signal.h` is standard C89, let's drop the guard. Also from the rest of the code.

`pthread.h` needs detection in autotools too.

These said, I'm also fine if we don't add this sample/test code to the test.

vszakats commented on tests/test_ssh2.c:
```diff
@@ -63,6 +92,24 @@ int main(int argc, char *argv[])
+    sigemptyset(&signal_handler_config.sa_mask);
+
+    if(sigaction(SIGPROF, &signal_handler_config, NULL) != 0) {
+        fprintf(stderr, "Could not install signal handler\n");
```
```suggestion
        fprintf(stderr, "Could not install signal handler.\n");
```

vszakats commented on tests/test_ssh2.c:
```diff
@@ -63,6 +92,24 @@ int main(int argc, char *argv[])
     }
 #endif
 
+#if defined(HAVE_PTHREAD_H) && defined(HAVE_SIGNAL_H)
```
```suggestion
#if defined(HAVE_PTHREAD_H)
```

vszakats commented on tests/test_ssh2.c:
```diff
@@ -63,6 +92,24 @@ int main(int argc, char *argv[])
+        .sa_flags = SA_RESTART | SA_SIGINFO,
+        .sa_sigaction = signal_handler_function
+    };
+    sigemptyset(&signal_handler_config.sa_mask);
```
`sigemptyset` is non-portable.

sodabrew replied:
Also OK to be non-portable for the same reason as below

vszakats commented on tests/test_ssh2.c:
```diff
@@ -63,6 +92,24 @@ int main(int argc, char *argv[])
+    };
+    sigemptyset(&signal_handler_config.sa_mask);
+
+    if(sigaction(SIGPROF, &signal_handler_config, NULL) != 0) {
```
`sigaction` also non-portable.

sodabrew replied:
SIGPROF is the whole reason for this PR. It is used in Linux to gather process status. Failing to handle SIGPROF results in SSH connection failure rather than allowing the application using the library to be profiled by an external measurement tool.

It is OK to be non-portable so long as it's properly wrapped to prevent breaking on systems that don't have SIGPROF.

Because all systems (that I am aware of) with `SIGPROF` also have `sigaction`, I would also see no reason not to use `sigaction` HTH!

vszakats replied:
We support more than just Linux. Also as I understand this is a causal example code committed to `tests`, but the essence of the feature is in `src`. CI runs fine now, but `pthread.h` detection is missing from autotools, and broken in CMake, so none of the example code is actually built.

Fixing autotools to have pthread detection might be one step forward, but it is complex. Even CMake is non-trivial if I look at how curl does these. We should also limit these detections to builds with tests enabled, otherwise it makes all builds slower. Only after implemeting these detection, could we see how the code compiles and runs in CI.

sodabrew commented:
Thanks for digging into this PR @vszakats! Let me know if there's anything you'd like me to work on, or if you've got it from here?

vszakats commented:
Thanks @sodabrew. This is now blocked by a non-portable signal functions and constants in `test_ssh2.c`. And pthread dependence. Also waiting for tests to finish. I'm siding with merging this, but without the test bits where all these problems are.

vszakats commented:
Thank you! Merged without the example code from tests.

vszakats commented:
For the record, here's the skipped part of the diff:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9b0ecca21..67c9fa8d7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -212,6 +212,7 @@ if(NOT WIN32)
   check_include_files("sys/un.h" HAVE_SYS_UN_H)
   check_include_files("arpa/inet.h" HAVE_ARPA_INET_H)  # example and tests
   check_include_files("netinet/in.h" HAVE_NETINET_IN_H)  # example and tests
+  check_include_files("pthread.h" HAVE_PTHREAD_H)  # tests
 endif()
 
 # CMake uses C syntax in check_symbol_exists() that generates a warning with
diff --git a/tests/test_ssh2.c b/tests/test_ssh2.c
index 2e6fa2382..a7f0f3743 100644
--- a/tests/test_ssh2.c
+++ b/tests/test_ssh2.c
@@ -24,6 +24,13 @@
 #include <stdio.h>
 #include <stdlib.h>  /* for getenv() */
 
+#if defined(HAVE_PTHREAD_H)
+#include <pthread.h>
+#include <signal.h>
+
+static pthread_t main_thread;
+#endif
+
 static const char *hostname = "127.0.0.1";
 static const int port_number = 4711;
 static const char *pubkey = "key_rsa.pub";
@@ -40,6 +47,28 @@ static void portable_sleep(unsigned int seconds)
 #endif
 }
 
+#if defined(HAVE_PTHREAD_H)
+static void *background_thread(void *unused)
+{
+    (void)unused;
+    fprintf(stderr, "Profiling thread sending SIGPROF every 1ms.\n");
+
+    do {
+        pthread_kill(main_thread, SIGPROF);
+        usleep(1000);
+    } while(1);
+
+    return NULL;
+}
+
+static void signal_handler_function(int a, siginfo_t *b, void *c)
+{
+    (void)a;
+    (void)b;
+    (void)c;
+}
+#endif
+
 int main(int argc, char *argv[])
 {
     uint32_t hostaddr;
@@ -63,6 +92,26 @@ int main(int argc, char *argv[])
     }
 #endif
 
+#if defined(HAVE_PTHREAD_H)
+    {
+        pthread_t thread;
+
+        struct sigaction signal_handler_config = {
+            .sa_flags = SA_RESTART | SA_SIGINFO,
+            .sa_sigaction = signal_handler_function
+        };
+        sigemptyset(&signal_handler_config.sa_mask);
+
+        if(sigaction(SIGPROF, &signal_handler_config, NULL) != 0) {
+            fprintf(stderr, "Could not install signal handler.\n");
+        }
+
+        main_thread = pthread_self();
+
+        pthread_create(&thread, NULL, background_thread, NULL);
+    }
+#endif
+
     (void)argc;
     (void)argv;
 
@@ -160,7 +209,7 @@ int main(int argc, char *argv[])
                rc != LIBSSH2_ERROR_KEY_EXCHANGE_FAILURE ||
 #endif
                ++retry > retries) {
-                break;
+                goto shutdown;
             }
             fprintf(stderr, "Retrying... %d / %d\n", retry, retries);
         } while(1);
```
